components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/MetricsPingScheduler.kt

```diff
@@ -6,77 +6,255 @@ package mozilla.components.service.glean.scheduler
+            else -> {
+                // We're overdue and don't want to wait until tomorrow.
+                0L
+            }
+        }
+    } /*
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun getMillisecondsUntilDueTime(
+        now: Calendar = getCalendarInstance(),
+        dueHourOfTheDay: Int = DUE_HOUR_OF_THE_DAY
+    ): Long {
+        val nowInMillis = now.timeInMillis
+        val dueTime = now
+        dueTime.set(Calendar.HOUR_OF_DAY, dueHourOfTheDay)
+        dueTime.set(Calendar.MINUTE, 0)
+        dueTime.set(Calendar.SECOND, 0)
+        dueTime.set(Calendar.MILLISECOND, 0)
+
+        val delay = dueTime.timeInMillis - nowInMillis
+        return when {
+            delay >= 0 -> {
+                // The `dueHourOfTheDay` is in the current calendar day.
+                // Return the computed delay.
+                delay
+            }
+            else -> {
+                // We're past the `dueHourOfTheDay` in the current calendar day.
+                dueTime.add(Calendar.DAY_OF_MONTH, 1)
+                dueTime.timeInMillis - nowInMillis
+            }
         }
+    }*/
 
-        val currentTime = System.currentTimeMillis()
+    /**
+     * TODO
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun isAfterDueTime(
+        now: Calendar = getCalendarInstance(),
+        dueHourOfTheDay: Int = DUE_HOUR_OF_THE_DAY
+    ): Boolean {
+        val nowInMillis = now.timeInMillis
+        val dueTime = getDueTimeForToday(now, dueHourOfTheDay)
+        return (dueTime.timeInMillis - nowInMillis) < 0
+    }
 
-        if (lastPingTime == 0L ||
-            currentTime - lastPingTime >= TimeUnit.HOURS.toMillis(PING_INTERVAL_HOURS)) {
-            return true
+    /**
+     * TODO
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun getDueTimeForToday(now: Calendar, dueHourOfTheDay: Int): Calendar {
+        val dueTime = now.clone() as Calendar
+        dueTime.set(Calendar.HOUR_OF_DAY, dueHourOfTheDay)
+        dueTime.set(Calendar.MINUTE, 0)
+        dueTime.set(Calendar.SECOND, 0)
+        dueTime.set(Calendar.MILLISECOND, 0)
+        return dueTime
+    }
+
+    /**
+     * TODO
+     */
+    fun startupCheck() {
+        val lastSentDate = getLastCollectedDate()
+
+        // We expect to cover 3 cases here:
+        //
+        // (1) - the ping was already collected the current calendar day; only schedule
+        //       one for collecting the next calendar day at the due time;
+        // (2) - the ping was NOT collected the current calendar day, and we're later than
+        //       the due time; collect the ping immediately;
+        // (3) - the ping was NOT collected the current calendar day, but we still have
+        //       some time to the due time; schedule for sending the current calendar day.
+
+        val alreadySentToday = DateUtils.isToday(lastSentDate.time)
+        if (alreadySentToday) {
```

Comment:
Minor point: Might be a good place for `when` block here, especially because of the early returns.

Reply:
Ah! Good point, this also helps with readability.